Source/WebGPU/WebGPU/Device.mm

```diff
@@ -151,13 +151,13 @@ static void captureFrame(id<MTLDevice> captureObject)
 {
 #if PLATFORM(MAC)
     auto devices = MTLCopyAllDevicesWithObserver(&m_deviceObserver, [weakThis = ThreadSafeWeakPtr { *this }](id<MTLDevice> device, MTLDeviceNotificationName) {
-        RefPtr<Device> strongThis = weakThis.get();
-        if (!strongThis)
+        RefPtr<Device> protectedThis = weakThis.get();
+        if (!protectedThis)
             return;
-        strongThis->instance().scheduleWork([strongThis = WTFMove(strongThis), device = device]() {
-            if (![strongThis->m_device isEqual:device])
+        protectedThis->instance().scheduleWork([protectedThis = WTFMove(protectedThis), device = device]() {
```

Comment:
Not your fault but I'm not convinced the WTFMove() is safe here since we're using protectedThis in the same instruction.

Reply:
indeed, this is undefined behaviour, more likely than not as use after move.

Reply:
I didn't author the code but it seems to rely on `strongThis->instance()` evaluating before the function arguments to the chained function `scheduleWork` are evaluated, which I think is guaranteed by C++17 and that would explain why this code which is evaluated every launch has not observed a failure.

But it seems like we should clean that up and not rely on such evaluation orders.

Reply:
> I didn't author the code but it seems to rely on `strongThis->instance()` evaluating before the function arguments to the chained function `scheduleWork` are evaluated, which I think is guaranteed by C++17

Yes, this will occur before scheduleWork is run, however this happens after the the lambda is constructed, and the capture run.

I just did a test in a debug build, and indeed it crashes with a nullderef. 
I used NativePromise API test for convenience.
```
    RefPtr<GenericPromise> promise;
    promise->whenSettled(queue, [queue, promise = WTFMove(promise)] {
        queue->beginShutdown();
        (void)promise;
    });
```

that gives:
```
(lldb) bt
* thread #1, queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=1, address=0xc)
  * frame #0: 0x0000000100078390 TestWTF`bool std::__1::__cxx_atomic_load[abi:v160006]<bool>(__a=0x000000000000000c, __order=seq_cst) at atomic:963:12
    frame #1: 0x0000000100078318 TestWTF`std::__1::__atomic_base<bool, false>::load[abi:v160006](this=0x000000000000000c, __m=seq_cst) const at atomic:1550:17
    frame #2: 0x0000000100078230 TestWTF`std::__1::__atomic_base<bool, false>::operator bool[abi:v160006](this=0x000000000000000c) const at atomic:1554:53
    frame #3: 0x0000000100078a4c TestWTF`WTF::ThreadSafeRefCountedBase::ref(this=0x0000000000000008) const at ThreadSafeRefCounted.h:58:9
    frame #4: 0x0000000100454058 TestWTF`WTF::Ref<WTF::NativePromise<void, int, 0u>, WTF::RawPtrTraits<WTF::NativePromise<void, int, 0u>>>::Ref(this=0x000000016fdfece8, object=0x0000000000000000) at Ref.h:67:16
    frame #5: 0x0000000100453ac4 TestWTF`WTF::Ref<WTF::NativePromise<void, int, 0u>, WTF::RawPtrTraits<WTF::NativePromise<void, int, 0u>>>::Ref(this=0x000000016fdfece8, object=0x0000000000000000) at Ref.h:66:5
    frame #6: 0x000000010045be7c TestWTF`WTF::NativePromise<void, int, 0u>::ThenCommand<WTF::NativePromise<void, int, 0u>::ThenCallback<false, void>>::ThenCommand(this=0x000000016fdfece8, promise=0x0000000000000000, thenCallback=0x000000016fdfec08, callSite=0x000000016fdfecc0) at NativePromise.h:814:15
    frame #7: 0x000000010045a500 TestWTF`WTF::NativePromise<void, int, 0u>::ThenCommand<WTF::NativePromise<void, int, 0u>::ThenCallback<false, void>>::ThenCommand(this=0x000000016fdfece8, promise=0x0000000000000000, thenCallback=0x000000016fdfec08, callSite=0x000000016fdfecc0) at NativePromise.h:817:9
    frame #8: 0x00000001003f96ec TestWTF`auto WTF::NativePromise<void, int, 0u>::whenSettled<WTF::Ref<TestWebKitAPI::WorkQueueWithShutdown, WTF::RawPtrTraits<TestWebKitAPI::WorkQueueWithShutdown>>, TestWebKitAPI::NativePromise_TakePhotoExample_Test::TestBody()::$_68>(this=0x0000000000000000, targetQueue=0x000000016fdfed28, settleFunction=0x000000016fdfecd8, callSite=0x000000016fdfecc0) at NativePromise.h:945:16
    frame #9: 0x00000001003f9544 TestWTF`TestWebKitAPI::NativePromise_TakePhotoExample_Test::TestBody(this=0x00006000006b2420) at NativePromise.cpp:1518:14
    frame #10: 0x0000000100a855e0 TestWTF`void testing::internal::HandleSehExceptionsInMethodIfSupported<testing::Test, void>(object=0x00006000006b2420, method=(actual=0x0000000100a38520 TestWTF`_ZN7testing4Test8TestBodyEv_vfpthunk_ at gtest.cc:2707:3), location="the test body") at gtest.cc:2607:10
    frame #11: 0x0000000100a3824c TestWTF`void testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>(object=0x00006000006b2420, method=(actual=0x0000000100a38520 TestWTF`_ZN7testing4Test8TestBodyEv_vfpthunk_ at gtest.cc:2707:3), location="the test body") at gtest.cc:2643:14
    frame #12: 0x0000000100a38180 TestWTF`testing::Test::Run(this=0x00006000006b2420) at gtest.cc:2682:5
    frame #13: 0x0000000100a39418 TestWTF`testing::TestInfo::Run(this=0x000000014af3ae20) at gtest.cc:2861:11
    frame #14: 0x0000000100a3a768 TestWTF`testing::TestSuite::Run(this=0x000000014af386e0) at gtest.cc:3015:28
    frame #15: 0x0000000100a48c08 TestWTF`testing::internal::UnitTestImpl::RunAllTests(this=0x000000014af04a10) at gtest.cc:5851:44
    frame #16: 0x0000000100a85f78 TestWTF`bool testing::internal::HandleSehExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(object=0x000000014af04a10, method=(actual=0x0000000100a4875c TestWTF`testing::internal::UnitTestImpl::RunAllTests() at gtest.cc:5742), location="auxiliary test code (environments or event listeners)") at gtest.cc:2607:10
    frame #17: 0x0000000100a484d0 TestWTF`bool testing::internal::HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(object=0x000000014af04a10, method=(actual=0x0000000100a4875c TestWTF`testing::internal::UnitTestImpl::RunAllTests() at gtest.cc:5742), location="auxiliary test code (environments or event listeners)") at gtest.cc:2643:14
    frame #18: 0x0000000100a483b4 TestWTF`testing::UnitTest::Run(this=0x0000000100f44958) at gtest.cc:5434:10
    frame #19: 0x00000001008096ec TestWTF`RUN_ALL_TESTS() at gtest.h:2471:46
    frame #20: 0x0000000100809678 TestWTF`TestWebKitAPI::TestsController::run(this=0x0000000100f44378, argc=1, argv=0x000000016fdff658) at TestsController.cpp:89:13
    frame #21: 0x0000000100a2a1bc TestWTF`main(argc=2, argv=0x000000016fdff658) at mainMac.mm:83:62
    frame #22: 0x00000001826c5140 dyld`start + 2360
```

Reply:
You're correct

https://en.cppreference.com/w/cpp/language/eval_order
"(since C++ 17) In a pointer-to-member expression E1.*E2 or E1->*E2, every value computation and side effect of E1 is sequenced before every value computation and side effect of E2 (unless the dynamic type of E1 does not contain the member to which E2 refers)."